src/Snail/Lexer.hs

```diff
@@ -1,50 +1,130 @@
+  sExpression,
+  sExpressions,
+
+  -- * Exported for testing
+  textLiteral,
 ) where
 
 import Data.Text (Text)
+import Data.Text qualified as Text
 import Data.Void
-import Text.Megaparsec
+import Snail.Characters
+import Text.Megaparsec hiding (token)
 import Text.Megaparsec.Char
 import Text.Megaparsec.Char.Lexer qualified as L
 
 -- | TODO: 'Void' is the error type but we should use an explicit error type
 type Parser = Parsec Void Text
 
--- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin with that prefix
+{- | Megaparsec's 'skipLineComment' takes a prefix and skips lines that begin
+ with that prefix
+-}
 skipLineComment :: Parser ()
 skipLineComment = L.skipLineComment ";"
 
--- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything in between
+{- | Megaparsec's 'skipBlockComment' takes prefix and suffix and skips anything
+ in between
+-}
 skipBlockComment :: Parser ()
 skipBlockComment = L.skipBlockCommentNested "#|" "|#"
 
--- | Generate a parser for whitespace in a language with 'skipLineComment' and 'skipBlockComment'
+{- | Generate a parser for whitespace in a language with 'skipLineComment' and
+ 'skipBlockComment'
+-}
 spaces :: Parser ()
 spaces = L.space space1 skipLineComment skipBlockComment
 
--- | A 'lexeme' is ...
-lexeme :: Parser a -> Parser a
-lexeme = L.lexeme spaces
-
 -- | Parse a 'Text' verbatim
 symbol :: Text -> Parser Text
 symbol = L.symbol spaces
 
--- | Parse an 'Integer' as a lexeme
-integer :: Parser Integer
-integer = lexeme L.decimal
-
--- | Parse a signed 'Integer'
-signedInteger :: Parser Integer
-signedInteger = L.signed spaces integer
-
 -- | Parse an S-Expression
 parens :: Parser a -> Parser a
 parens = between (symbol "(") (symbol ")")
+
+-- | The list of valid token characters, note that we allow invalid tokens at this point
+validCharacter :: Parser Char
+validCharacter =
+  oneOf
+    ( initialCharacter
+        <> specialInitialCharacter
+        <> digitCharacter
+        <> specialSubsequentCharacter
+    )
+
+{- | A possibly empty tree of s-expressions
+
+ Technically,
+ @
+ Token (SourcePos {..}, "hello")
+ @
+
+ isn't a valid s-expression. This is,
+
+ @
+ SExpression [Token (SourcePos {..}, "hello")]
+ @
+
+ and this is also valid,
+
+ @
+ SExpression []
+ @
+
+ The 'Data.Tree.Tree' type in containers is non-empty which isn't exactly what we are looking for
+
+ TODO: can we have a smart constructor only?
+ TODO: add TextLiteral for literal strings (check r5rs for literal string syntax, I think it is "..." and not '...')
+-}
+data SExpression
+  = Token (SourcePos, Text)
+  | TextLiteral (SourcePos, Text)
+  | SExpression [SExpression]
+  deriving (Eq, Show)
+
+-- | ...
+term :: Parser SExpression
+term = token <|> textLiteral <|> sExpression
+
+-- | ...
+token :: Parser SExpression
+token = do
+  sourcePosition <- getSourcePos
+  token' <- some validCharacter
+  pure $ Token (sourcePosition, Text.pack token')
+
+-- | ...
```

Comment:
I'd just delete the empty comments.

Reply:
They are sort of an invitation to start writing comments. I plan to start writing them sooner than later fwiw.